gradle-plugin/src/main/groovy/org/groovymc/modsdotgroovy/gradle/tasks/AbstractMDGConvertTask.groovy

```diff
@@ -0,0 +1,172 @@
+
+import javax.inject.Inject
+import java.nio.file.Files
+
+@CacheableTask
+@CompileStatic
+abstract class AbstractMDGConvertTask extends DefaultTask {
+    private static final CompilerConfiguration MDG_COMPILER_CONFIG = new CompilerConfiguration().tap {
+        targetBytecode = JDK17
+        optimizationOptions['indy'] = true
+    }
+
+    @InputFile
+    @NormalizeLineEndings
+    @PathSensitive(PathSensitivity.NONE)
+    abstract RegularFileProperty getInput()
+
+    @Optional
+    @OutputFile
+    abstract RegularFileProperty getOutput()
+
+    @Input
+    abstract Property<String> getOutputName()
+
+    @CompileClasspath
+    abstract ConfigurableFileCollection getMdgRuntimeFiles()
+
+    @Input
+    @Optional
+    abstract MapProperty<String, Object> getBuildProperties()
+
+    @Input
+    @Optional
+    abstract SetProperty<String> getEnvironmentBlacklist()
+
+    @Input
+    @Optional
+    abstract Property<Platform> getPlatform()
+
+    @Input
+    abstract Property<String> getProjectVersion()
+
+    @Input
+    abstract Property<String> getProjectGroup()
+
+    /**
+     * A file containing platform-specific details, such as the Minecraft version and loader version.
+     * <p>Generated by a task that extends {@link AbstractGatherPlatformDetailsTask}</p>
+     */
+    @InputFile
+    @NormalizeLineEndings
+    @PathSensitive(PathSensitivity.NONE)
+    abstract RegularFileProperty getPlatformDetailsFile()
+
+    @Inject
+    protected abstract ProjectLayout getProjectLayout()
+
+    AbstractMDGConvertTask() {
+        input.convention(projectLayout.projectDirectory.file('src/main/resources/mods.groovy'))
+        // default to e.g. build/modsDotGroovyToToml/mods.toml
+        output.convention(projectLayout.buildDirectory.dir('generated/modsDotGroovy/' + name.replaceFirst('ConvertTo', 'modsDotGroovyTo')).map((Directory dir) -> dir.file(outputName.get())))
+
+        environmentBlacklist.convention(project.provider(() -> mdgExtension.environmentBlacklist.get()))
+        buildProperties.convention(project.provider(() -> filterBuildProperties(project.extensions.extraProperties.properties, environmentBlacklist.get())))
+        platform.convention(project.provider(() -> {
+            final @Nullable Set<Platform> platforms = mdgExtension.platforms?.get()?.keySet()
+            return platforms === null || platforms.isEmpty() ? Platform.FORGE : platforms.first()
+        }))
+        projectVersion.convention(project.provider(() -> project.version.toString()))
+        projectGroup.convention(project.provider(() -> project.group.toString()))
+        platformDetailsFile.convention(projectLayout.buildDirectory.dir("generated/modsDotGroovy/gather${platform.get()}PlatformDetails").map((Directory dir) -> dir.file('mdgPlatform.json')))
+    }
+
+    protected abstract String writeData(Map data)
+
+    protected static Map<String, Object> filterBuildProperties(final Map<String, Object> buildProperties, final Set<String> blacklist) {
```

Comment:
I would move from a blacklist to a whitelist given the rather easy way one can expose secrets stored in the user-level gradle.properties with a blacklist.

Reply:
The blacklist system is based on the property name containing certain words rather than an exact match - this means any property name containing "secret", "pass", etc are automatically excluded to help avoid this issue.